### ISSUE TITLE
Question about tracking events in Application Insights

### ISSUE THREAD

**Issue by erndob** [Microsoft/ApplicationInsights-dotnet-logging#254]
I'm trying to port my TelemetryClient usage to ILogger with ApplicationInsightsProvider.

I have a lot of TrackEvent logging. 

I see that the current implementation, always uses  TraceTelemetry, and never uses EventTelemetry, yet it does populate the EventName property if the EventId is provided. 

My question is, is there any actual difference from ApplicationInsights perspective if I change all EventTelemetry to TraceTelemetry, but with same EventName property?

I guess it would be possible to code the case, where message is empty but the EventId object has a name, and then simply instead of creating TraceTelemetry we could create EventTelemetry. Is that something we could add?

**Comment by RamjotSingh** [comment]
TraceTelemetry goes to traces table on ApplicationInsights, whereas EventTelemetry goes to customEvents. Traces table is optimized for viewing traces (has the right properties for it to query).

Now the choice of EventTelemetry vs TraceTelemetry is a different one. EventTelemetry is used to track specific events e.g. every time you make a cache call you might to record if the cache call resulted into a miss or not, whereas Trace would be something like recording the thumbprint of the certificate you were trying to read from disk storage. There are times where line between Event and Trace can be blurred.

ILogger records to Traces and Exceptions tables only.

Hope this answers you question.

**Comment by aKzenT** [comment]
Would you accept a contribution that allows to configure whether ApplicationInsightsLogger records as Traces or as Events?

We want to use ILogger for our logging, but we want some messages tracked as Events, other as Traces. If we can configure ApplicationInsightsLogger, we could register two instances and use filters to send events to one instance and traces to other instance.

**Comment by cijothomas** [comment]
@aKzenT Ilogger logs naturally map to Traces in application insights model. All other logging adapters create TraceTelemetry only, so having an option to product Events is likely confusing.

Can you instead directly use TelemetryClient.TrackEvent when events are appropriate? Or are you planning to avoid taking direct application insights dependency, and use ILogger only everywhere?

**Comment by aKzenT** [comment]
@cijothomas Yes, we want to use ILogger everywhere and try to avoid having a direct dependency on application insights just for "events". It feels naturally to use the same logging infrastructure for both. We have used Serilog in the past with application insights as a sink and Serilog does indeed offer the choice between trace and event:

https://github.com/serilog/serilog-sinks-applicationinsights/tree/master/src/Serilog.Sinks.ApplicationInsights/Sinks/ApplicationInsights/TelemetryConverters

**Comment by johnib** [comment]
Can you update on this? Would appreciate if you accept the contribution that was proposed by @erndob

**Comment by SergeyKanzhelev** [comment]
Converter like in SeriLog can be implemented as a TelemetryProcessor. Just take all `TraceTelemetry` and convert to `EventTelemetry`. If you are thinking of ILogger that accept a special custom type of logs and understands it's semantic - perhaps a custom ILogger can be developer. It shouldn't be more than a couple pages of code as I understand. Mostly just mapping of properties.

Am I missing something here?

**Comment by johnib** [comment]
Worst case, everyone can fork this repo and add their customization to it and use it by their own.

But as mentioned, IMO this is the worst case, for so many reasons.

The fact here is that we are not adding some customization that we pulled out of nowhere, this is the kind of customization that compliments the product for which this repo was even invented.

Events is a notion that both the ILogger interface and Application insights as a product - have, it only makes sense to make this AI logger have this notion as well, and only for those who wants it. 

Same argument holds for Exeptions, where both ILogger and Application insights know and offer different treatment, than just a simple trace. 
However, in this case, the writer of this lib found it is justified to allow treating exceptions differently (if desired). 

Why not for events then?

**Comment by amigup** [comment]
@cijothomas - can you please provide the update on this thread? I concur with @aKzenT that there should be one abstraction to log both trace and event as both are first class citizen of Application insights.

**Comment by johnib** [comment]
@Cijo Thomas<mailto:cithomas@microsoft.com>  gentle ping

Jonathan
________________________________

**Comment by cijothomas** [comment]
Haven't got the time to review this further. 
@SergeyKanzhelev do you have any suggestion here?

**Comment by cijothomas** [comment]
@Dmitry-Matveev also

**Comment by Dmitry-Matveev** [comment]
- If EventId.Name is not empty && message is empty - use EventTelemetry and EventId.Name;
- If EventId.Name is not empty && message is not empty - use TraceTelemetry and no name;
- If EventId.Name is not empty && message is not empty && Exception is not empty - use Exception and nothing else;

Those are only cases for EventId.Name being non-empty, we can multiply by having other combinations. This is slightly more confusing when applied instead of "If it has Exception object, It's exception, otherwise Trace".

Having the decision tree above may increase support cost of this SDK due to "my traces being lost" cases where they hide in Event Table or "my events being lost" while they hide in the Traces table.

The current implementation of all logging adapters shipped by AI team only considers Trace or Exception types, while Serilog adapter is a community contribution and is not in the support map of SDKs.

I'm not stating that we cannot do such a change, but that it's out of what this repo educated our customers about for years (thus, confusing), and it will cost us more in support than it will achieve (thus, low ROI). It's not as simple as "this repo is about logging, so we should add events".

**Comment by johnib** [comment]
Yeah, looks too complex I agree.
I could be simpler though:

1. Have a config flag "shouldTrackCustomEvent" -- which, if enabled, sends custom events whenever the EventId.Name is populated, and composes the custom dimensions as it is done for traces as well

2. If the config flag is disabled, then I do not understand what's the meaning of the EventId.Name anyway -- I do not recall it is being used -- could be wrong.

This way, everyone who is used to how the SDK operates today will have no "breaking" change.
I expect the support cost to remain as they are today.

What do you think @Dmitry-Matveev ?

Thanks
